### PR TITLE
Enable retries on some 500 codes on reads

### PR DESCRIPTION
**Pull request by damccorm**
This should enable https://github.com/Microsoft/azure-devops-node-api/issues/288 to be done more easily

### CODE REVIEW

File: lib/HttpClient.ts [hunk 1]
```diff
@@ -186,48 +198,62 @@ export class HttpClient implements ifm.IHttpClient {
         }
 
         let info: RequestInfo = this._prepareRequest(verb, requestUrl, headers);
-        let response: HttpClientResponse = await this.requestRaw(info, data);
 
-        // Check if it's an authentication challenge
-        if (response && response.message && response.message.statusCode === HttpCodes.Unauthorized) {
-            let authenticationHandler: ifm.IRequestHandler;
+        // Only perform retries on reads since writes may not be idempotent.
+        let numTries = (this._allowRetries && HttpWriteOptions.indexOf(verb) == -1) ? this._maxRetries + 1 : 1;
```
Comment (stephenmichaelf) [inline lib/HttpClient.ts]: numTriesRemaining?

Comment (stephenmichaelf) [inline lib/HttpClient.ts]: numTriesRemaining?

File: lib/HttpClient.ts [hunk 2]
```diff
@@ -186,48 +198,62 @@ export class HttpClient implements ifm.IHttpClient {
...
-                break;
+                // let's make the request with the new redirectUrl
+                info = this._prepareRequest(verb, redirectUrl, headers);
+                response = await this.requestRaw(info, data);
+                redirectsRemaining--;
             }
 
-            // we need to finish reading the response before reassigning response
-            // which will leak the open socket.
-            await response.readBody();
```
Comment (stephenmichaelf) [inline lib/HttpClient.ts]: I think we need to keep this.

File: lib/HttpClient.ts [hunk 3]
```diff
@@ -38,6 +38,8 @@ export enum HttpCodes {
 }
 
 const HttpRedirectCodes: number[] = [HttpCodes.MovedPermanently, HttpCodes.ResourceMoved, HttpCodes.SeeOther, HttpCodes.TemporaryRedirect, HttpCodes.PermanentRedirect];
+const HttpRetryCodes: number[] = [HttpCodes.BadGateway, HttpCodes.ServiceUnavailable, HttpCodes.GatewayTimeout];
+const HttpWriteOptions: string[] = ['DELETE', 'POST', 'PUT', 'PATCH'];
```
Comment (stephenmichaelf) [inline lib/HttpClient.ts]: Do we need to expose this?

File: lib/HttpClient.ts (hunk 2 above)
Comment (damccorm) [reply lib/HttpClient.ts]: Yeah, its still there, the diff is just displaying differently. Its on line 243 and occurs in the same place relative to the original code.

File: lib/HttpClient.ts (hunk 3 above)
Comment (damccorm) [reply lib/HttpClient.ts]: I don't think so - what would it be used for? I'd rather expose less if we don't have a clear use case

**Comment by stephenmichaelf** [comment]
I think we have tests for redirect, can we add some for this too?

**Comment by damccorm** [comment]
Good call, added a test which actually caught an issue!

File: lib/HttpClient.ts [hunk 4]
```diff
@@ -186,48 +198,62 @@ export class HttpClient implements ifm.IHttpClient {
...
-            let authenticationHandler: ifm.IRequestHandler;
+        // Only perform retries on reads since writes may not be idempotent.
+        let numTriesRemaining = (this._allowRetries && HttpWriteOptions.indexOf(verb) == -1) ? this._maxRetries + 1 : 1;
 
-            for (let i = 0; i < this.handlers.length; i++) {
-                if (this.handlers[i].canHandleAuthentication(response)) {
-                    authenticationHandler = this.handlers[i];
-                    break;
+        let response: HttpClientResponse;
+        while (numTriesRemaining > 0) {
```
Comment (stephenmichaelf) [inline lib/HttpClient.ts]: We should add a sleep/exponential backoff.

File: lib/HttpClient.ts (hunk 3 above)
Comment (stephenmichaelf) [inline lib/HttpClient.ts]: Delete is probably fine. Can we change this to opt in list instead of opt out?

File: lib/HttpClient.ts [hunk 5]
```diff
@@ -38,6 +38,8 @@ export enum HttpCodes {
 }
 
 const HttpRedirectCodes: number[] = [HttpCodes.MovedPermanently, HttpCodes.ResourceMoved, HttpCodes.SeeOther, HttpCodes.TemporaryRedirect, HttpCodes.PermanentRedirect];
+const HttpRetryCodes: number[] = [HttpCodes.BadGateway, HttpCodes.ServiceUnavailable, HttpCodes.GatewayTimeout];
```
Comment (stephenmichaelf) [inline lib/HttpClient.ts]: HttpResponseRetryCodes.

File: lib/HttpClient.ts (hunk 3 above)
Comment (stephenmichaelf) [reply lib/HttpClient.ts]: Maybe rename to RetryableHttpVerbs?

File: test/units/httptests.ts [hunk 6]
```diff
@@ -422,4 +422,19 @@ describe('Http Tests with keepAlive', function () {
         let res: httpm.HttpClientResponse = await _http.options('http://microsoft.com');
         assert(res.message.statusCode == 200, "status code should be 200");
     });
+
+    it('handles retries correctly', async() => {
+        _http = new httpm.HttpClient('typed-test-client-tests', null, {allowRetries: true, maxRetries: 1});
```
Comment (stephenmichaelf) [inline test/units/httptests.ts]: Lets do 4 or 5 so we make sure there are no off by one issues at a value of 1.

File: lib/HttpClient.ts [hunk 7]
```diff
@@ -186,48 +198,62 @@ export class HttpClient implements ifm.IHttpClient {
...
+                    break;
+                }
 
-            const redirectUrl: any = response.message.headers["location"];
-            if (!redirectUrl) {
-                // if there's no location to redirect to, we won't
-                break;
+                // we need to finish reading the response before reassigning response
+                // which will leak the open socket.
+                await response.readBody();
```
Comment (stephenmichaelf) [inline lib/HttpClient.ts]: I think we need to also do this in the outer retry loop?

**Comment by stephenmichaelf** [comment]
Looks like tests are failing.

**Comment by damccorm** [comment]
From offline discussion, this is a separate known issue with tests timing out and can be ignored.

**Comment by damccorm** [comment]
Accidentally merged this, has since been reverted and moved to #137

**Pull request merged**